ostd/src/task/scheduler/mod.rs

```diff
@@ -0,0 +1,292 @@
+// SPDX-License-Identifier: MPL-2.0
+
+//! Scheduling subsystem (in-OSTD part).
+//!
+//! This module defines what OSTD expects from a scheduling implementation
+//! and provides useful functions for controlling the execution flow.
+#![allow(dead_code)]
+
+mod fifo_scheduler;
+
+pub use fifo_scheduler::FifoScheduler;
+use spin::Once;
+
+use super::{processor, task::Task, TaskStatus};
+use crate::{
+    arch::{self, timer},
+    cpu::this_cpu,
+    prelude::*,
+};
+
+static SCHEDULER: Once<&'static dyn Scheduler<Task>> = Once::new();
+
+static SHOULD_PREEMPT: ShouldPreemptFlag = ShouldPreemptFlag::new();
+
+/// Injects a scheduler implementation into framework.
+///
+/// This function can only be called once and must be called during the initialization of kernel.
+pub fn inject_scheduler(scheduler: &'static dyn Scheduler<Task>) {
+    SCHEDULER.call_once(|| scheduler);
+
+    timer::register_callback(|| {
+        SCHEDULER.get().unwrap().local_mut_rq_with(&mut |local_rq| {
+            if local_rq.update_current(UpdateFlags::Tick) {
+                SHOULD_PREEMPT.set();
+            }
+        })
+    });
+}
+
+/// Enqueues a newly built task.
+///
+/// Note that the new task is not guranteed to run at once.
+pub(super) fn run_new_task(runnable: Arc<Task>) {
+    // FIXME: remove this check for `SCHEDULER`.
+    // Currently OSTD cannot know whether its user has injected a scheduler.
+    if !SCHEDULER.is_completed() {
+        let fifo_scheduler = Box::new(FifoScheduler::default());
+        let static_scheduler = Box::leak(fifo_scheduler);
+        inject_scheduler(static_scheduler);
+    }
+
+    let cpu_id_option = SCHEDULER
+        .get()
+        .unwrap()
+        .enqueue(runnable, EnqueueFlags::Spawn);
+    if cpu_id_option.is_some() {
+        let cpu_id = cpu_id_option.unwrap();
+        // FIXME: send ipi to set remote CPU's `SHOULD_PREEMPT` if needed.
+        if cpu_id == this_cpu() {
+            SHOULD_PREEMPT.set();
+        }
+    }
+
+    preempt();
+}
+
+/// Dequeues the current task from its runqueue.
+///
+/// This should only be called if the current is to exit.
+pub(super) fn exit_current() {
+    reschedule(&mut |local_rq: &mut (dyn LocalRunQueue + '_)| {
+        if let Some(next) = local_rq.pick_next_current() {
+            ReschedAction::SwitchTo(next)
+        } else {
+            ReschedAction::Retry
+        }
+    })
+}
+
+/// Makes the current task sleeping (scheduler part).
+pub(crate) fn sleep_current() {
+    reschedule(&mut |local_rq: &mut dyn LocalRunQueue| {
+        local_rq.update_current(UpdateFlags::Wait);
+
+        if let Some(next) = local_rq.pick_next_current() {
+            ReschedAction::SwitchTo(next)
+        } else {
+            ReschedAction::DoNothing
```

Comment:
Doing nothing means go back to the loop in `do_wait` and check again if the current has been woken, which seems reasonable to me when we cannot find a candidate for next current.

Reply:
Sorry I don't get it. I'm afraid `do_wait` will never return unless there is a `wake_up` that sets the `has_woken` flag therefore no sleeping task can escape and go on running.

Our implementation is like the code below,
```rust
fn do_wait(&self) {
    let has_woken = &self.has_woken;
    while !has_woken.swap(false, Ordering::Acquire) {
        scheduler::park_current(has_woken);
    }
}
```
which is somewhat like a paradigm below,
```rust
fn action() {
    while cond() {
        try_action();
    }
}
``` 

However we can surely return `ReschedAction::Retry` here and let the thread spin in the loop of `reschedule` instead, which makes nearly no difference in functionality since we also have a `has_woken` check in `park_current`. Though I think going back to the loop in `do_wait` makes more sense because that's probably what the loop there is for.